modules/plugin/jdbc/jdbc-oracle/src/main/java/org/geotools/data/oracle/OracleDialect.java

```diff
@@ -617,23 +622,30 @@ OracleConnection unwrapConnection( Connection cx ) throws SQLException {
         }
         
         try {
-            // first lookup ever? (we have UNWRAPPER_NOT_FOUND as a sentinel for a lookup that
-            // will not work (we assume the datasource will always return connections we can
-            // unwrap, or never).
-            if (uw == null) {
-                UnWrapper unwrapper = DataSourceFinder.getUnWrapper(cx);
+            // Unwrap the connection multiple levels as necessary to get at the underlying
+            // OracleConnection. Maintain a map of UnWrappers to avoid searching
+            // the registry every time we need to unwrap.
+            Connection testCon = cx;
+            do {
+                UnWrapper unwrapper = uwMap.get(testCon.getClass());
                 if (unwrapper == null) {
-                    uw = UNWRAPPER_NOT_FOUND;
-                } else {
-                    uw = unwrapper;
+                    unwrapper = DataSourceFinder.getUnWrapper(testCon);
+                    if (unwrapper == null) {
+                        unwrapper = UNWRAPPER_NOT_FOUND;
+                    }
+                    uwMap.put(testCon.getClass(), unwrapper);
                 }
-            }
-            if (uw != null && uw != UNWRAPPER_NOT_FOUND) {
-                Connection uwcx = uw.unwrap( cx );
-                if ( uwcx != null && uwcx instanceof OracleConnection ) {
-                    return (OracleConnection) uwcx;
+                if (unwrapper == UNWRAPPER_NOT_FOUND) {
+                    // give up and try legacy approaches below
+                    break;
                 }
-            } else if (cx instanceof Wrapper) {
+                testCon = unwrapper.unwrap(testCon);
+                if (testCon instanceof OracleConnection) {
+                    return (OracleConnection) testCon;
+                }
+            } while (testCon != null);
```

Comment:
I'm somewhat worried about this one, if we stumble into a badly written un-wrapper (one that in some case returns the connection its given) we might never get out of the loop.
We might want to guard against that too